File: src-tauri/crates/app/src/commands/tasks.rs
//! Commands over the running-task registry (issue #601).
//!
//! Thin on purpose: the registry in [`crate::tasks`] owns the state and
//! the routing, and these two just expose it. The status bar subscribes
//! to `tasks:changed` and calls [`list_tasks`] once at mount — in that
//! order, because Tauri does not replay an event to a listener that
//! registered a moment too late.

use crate::{
    error::AppResult,
    tasks::{TaskRegistry, TaskSnapshot},
};
use std::sync::Arc;

#[tauri::command]
pub async fn list_tasks(
    registry: tauri::State<'_, Arc<TaskRegistry>>,
) -> AppResult<Vec<TaskSnapshot>> {
    Ok(registry.snapshot())
}

/// Ask a task to stop, through whatever mechanism it registered.
///
/// Returns `false` when there was nothing to ask: the task finished
/// between the render and the click, or it is already stopping, or it
/// never offered a way out. All three are ordinary — a cancel button
/// races the task it cancels by construction — so none of them is an
/// error.
#[tauri::command]
pub async fn cancel_task(
    registry: tauri::State<'_, Arc<TaskRegistry>>,
    id: u64,
) -> AppResult<bool> {
    Ok(registry.cancel(id))
}

File: src/lib/tauri/tasks.ts
import { invoke } from "@tauri-apps/api/core";

/** Event carrying the whole list whenever it changes. */
export const TASKS_CHANGED = "tasks:changed";

/**
 * One running long operation (issue #601).
 *
 * Mirrors `tasks::TaskSnapshot`. `kind` is a stable slug the UI turns
 * into localized copy — treat it like an event name, not a label.
 */
export interface TaskSnapshot {
  id: number;
  kind: string;
  /** A path or a title. Deliberately not localized. */
  detail: string | null;
  current: number;
  /** `0` means indeterminate. */
  total: number;
  cancellable: boolean;
  /** A stop has been asked for and the task has not exited yet. */
  cancelling: boolean;
}

export function listTasks(): Promise<TaskSnapshot[]> {
  return invoke<TaskSnapshot[]>("list_tasks");
}

/**
 * Ask a task to stop, through whatever mechanism it registered.
 *
 * `false` means there was nothing to ask — it finished between the
 * render and the click, it is already stopping, or it never offered a
 * way out. A cancel button races the task it cancels by construction,
 * so none of those is an error worth showing.
 */
export function cancelTask(id: number): Promise<boolean> {
  return invoke<boolean>("cancel_task", { id });
}
